designs/NuGet-RemoteFileService.md

```diff
@@ -0,0 +1,70 @@
+
+# Spec Name
+
+* Status: In Review
+* Author(s): [Rob Relyea](https://github.com/rrelyea)
+* Issue: [10151](https://github.com/NuGet/Home/issues/10151) INuGetRemoteFileService - Fetch Images and embedded licenses for codespaces and normal execution
+
+## Problem Background
+
+With Codespaces support added to NuGet client, our code needs to be able to handle all packages and communication with feeds happening on a server node. This is architected as async method calls to services that are remoted. In most cases, we also have chosen to have 1 code path, for normal execution and remoted execution. These services are still used in the normal execution, however, the service runs in the same process, as opposed to being remoted across the network.
+
+Package Icons can either come from:
+- a package feeds search service
+- (old style) or a URI to an arbitrary place on the web
+- embedded in a package
+
+Licenses can also come from a few places, including a license file embedded in a package.
+
+## Who are the customers
+
+All PM UI customers.
+
+## Goals
+
+Support remote and normal fetching of icons and licenses.
+
+## Non-Goals
+
+## Solution Overview
+
+### Fetching Files
+Defined a new interface for this service:
+
+namespace NuGet.VisualStudio.Internal.Contracts
+{
+    public interface INuGetRemoteFileService : IDisposable
+    {
+        ValueTask<Stream?> GetPackageIconAsync(PackageIdentity packageIdentity, CancellationToken cancellationToken);
+        ValueTask<Stream?> GetEmbeddedLicenseAsync(PackageIdentity packageIdentity, CancellationToken cancellationToken);

```

Comment:
1) don't think this API is where we should provide validation of package contents. Package quality like that should be enforced at several other times (pack/push/publish to nuget.org/etc...)

2) i'll think through the question about failing http(s) calls...

Reply:
about 2) i'm assuming the http(s) failing call you are talking about is the remoted call to the server. is that generally handled by Wrapper class, which i don't have in this work currently? perhaps we have a call to talk this through.

Reply:
we discussed. Rob is going to go dig into more of the license code paths.
(note: NuGetProjectManagerService has some examples of remote exceptions in some of the methods - try/catch)

Reply:
SearchObject.cs .. along with search metadata and icon info, license info might be stored here on the service side

if packagePath != null && licenseMetadata != null
	store license uri in PackageIdentityToUri memoryCache


on the client side:  PackageLicenseUtilities.cs
   looks like this will create three different data model objects...based on metadata from nuspec file
	licenseText -- URL
        LicenseText -- expression
	LicenseFileText -- embedded license file 


These two code paths are in click handlers that are only attached to LicenseFileText objects shown in UI via databinding:

packagemetadatacontrol.xaml.cs
   void ViewLicense_Click(sender, e)
       calls PackageLicenseUtilities.GetEmbeddedLicenseAsync

LicenseAcceptanceWindow.xaml.cs might call:
  licensefiletext.cs -- LoadLicenseFile() which loads an embedded license

Reply:
believe if there was a communication failure, an exception would be raised, and vs would crash. i've not added any other exception handling here. believe we are in a good place.
happy to talk through here or in code PR.